realm/app-manager/src/launcher/oci.rs

```
@@ -0,0 +1,351 @@
+        Self { launcher_config }
+    }
+
+    fn create_oci_client(&self, im_url: &str) -> Result<Client> {
+        let config = Config::builder().host(im_url.to_owned());
+
+        let oci_client = match &self.launcher_config {
+            OciLauncherConfig::NoTLS => Client::from_config(config.no_tls()),
+
+            OciLauncherConfig::RusTLS { root_ca } => Client::from_config(config.rustls_no_auth(
+                load_root_cert_store(root_ca).map_err(OciLauncherError::RootCaLoading)?,
+            )),
+
+            OciLauncherConfig::RaTLS {
+                root_ca,
+                token_resolver,
+            } => {
+                let token_resolver = match token_resolver {
+                    TokenResolver::FromFile(path) => TokenFromFile::from_path(path)
+                        .map_err(OciLauncherError::TokenReadingError)?,
+                };
+
+                let cert_resolver =
+                    RaTlsCertResolver::from_token_resolver(Arc::new(token_resolver))
+                        .map_err(OciLauncherError::RaTlsCertResolverCreation)?;
+
+                Client::from_config(config.ratls(
+                    load_root_cert_store(root_ca).map_err(OciLauncherError::RootCaLoading)?,
+                    Arc::new(cert_resolver),
+                ))
+            }
+        };
+
+        Ok(oci_client.map_err(OciLauncherError::OciClientCreation)?)
+    }
+
+    async fn try_read_metadata(&self, path: &Path) -> Option<Metadata> {
+        read_to_string(path.join("metadata.json"))
+            .await
+            .map(json_load)
+            .ok()
+            .transpose()
+            .ok()
+            .flatten()
+    }
+
+    async fn save_metadata(&self, path: &Path, metadata: Metadata) -> Result<()> {
+        write_to_file(path.join("metadata.json"), json_dump(metadata)?).await?;
+        Ok(())
+    }
+
+    async fn move_unpack_to_img(&self, unpack: &Path, img: &Path) -> Result<()> {
+        let _ = rmrf(img).await;
```

Minor: maybe we should add TRACE log with the result than ignore it totally?

---

Ok, I added trace as this call to rmrf will fail during provisioning application onto an empty disk. This is only useful during update. Therefore, I ignored the error message.